Source/Services.Clients/ReverseCallClient.cs

```diff
@@ -244,121 +159,293 @@ protected virtual void Dispose(bool disposing)
+            _clientToServer = connection.RequestStream;
+            _serverToClient = connection.ResponseStream;
+        }
+
+        Task SendConnectArguments(TConnectArguments connectArguments, CancellationToken cancellationToken)
+        {
+            _logger.SendingConnectArguments(typeof(TConnectArguments), _pingInterval);
+
+            var callContext = new ReverseCallArgumentsContext
+            {
+                ExecutionContext = _executionContextManager.Current.ToProtobuf(),
+                PingInterval = Duration.FromTimeSpan(_pingInterval)
+            };
+            _protocol.SetConnectArgumentsContext(callContext, connectArguments);
+            var message = new TClientMessage();
+            _protocol.SetConnectArguments(connectArguments, message);
+
+            return WriteMessage(message, cancellationToken);
+        }
+
+        async Task<bool> ReceiveConnectResponse(CancellationToken cancellationToken)
         {
-            await _writeResponseSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
             try
+            {
+                var stopwatch = Stopwatch.StartNew();
+                _logger.ReceivingConnectResponse(typeof(TConnectResponse));
+
+                using var keepalive = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                keepalive.CancelAfter(_pingInterval.Multiply(3));
+
+                if (await ReadMessageFromServerWhileRespondingToPings(keepalive).ConfigureAwait(false))
+                {
+                    var response = _protocol.GetConnectResponse(_serverToClient.Current);
+                    if (response == default)
+                    {
+                        _logger.DidNotReceiveConnectResponse(typeof(TConnectResponse), "Message did not contain a connect response");
+                        return false;
+                    }
+
+                    stopwatch.Stop();
+                    _logger.ReceivedConnectResponse(typeof(TConnectResponse), stopwatch.Elapsed);
+                    _metrics.AddToTotalWaitForConnectResponseTime(stopwatch.Elapsed);
+
+                    ConnectResponse = response;
+                    _connectionEstablished = true;
+                    return true;
+                }
+
+                _logger.DidNotReceiveConnectResponse(typeof(TConnectResponse), "No message received from server");
+                return false;
+            }
+            catch (RpcException ex) when (ex.StatusCode == StatusCode.Cancelled)
             {
                 if (cancellationToken.IsCancellationRequested)
                 {
-                    _logger.LogDebug("Reverse Call Client was cancelled before it could respond with pong");
-                    return;
+                    _logger.DidNotReceiveConnectResponse(typeof(TConnectResponse), "Connection was closed by the client");
+                }
+                else
+                {
```

Comment:
Would be nice to log the ping timeouts too during connecting, like in the `Handle()`.

```suggestion
                }
                else if (keepalive.IsCancellationRequested)
                {
	                  _metrics.IncrementTotalPingTimeouts();
	                  _logger.PingTimedOut();
                }
                else
                {
```